hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// TODO: This is conceptually kinda close to ReferenceGraph except more powerful
+
+// TypeWalkerRemoveType is a special TypeName that informs the type walker to remove the property containing this TypeName
+// entirely.
+var TypeWalkerRemoveType = MakeTypeName(LocalPackageReference{}, "TypeWalkerRemoveProperty")
+
+// TODO: it's awkward to have so much configuration on this thing (3 separate funcs that apply at different places in the walking process?)
```

Comment:
Bringing this up in code review.

If anybody has a cleaner way to make this thing less configurable while maintaining the power I'm 100% open to it. I'm basically trying to make a "super-visitor" that does a depth-first walk of a whole graph of types (including following the references between types). So the result of `Walk` is going to be the `Types` collection that is reachable from the root definition you provided, with members of that collection modified by the appropriate rules.

The two that feel most awkward to me are `MakeContextFunc` and `AfterVisitFunc`.
`AfterVisitFunc` for example is really only adding a _tiny_ bit on top of the visitor anyway, as it's allowing a modification of the `TypeDefinition` produced after the visitor has been called (which effectively boils down to the extra ability to fuss with the name of the type). It almost feels like this could be embedded into the visitor somehow and we could then remove this extra bit of configuration... but I poked at it and couldn't come up with a clean way.

Reply:
Those 2 are in contrast to `WalkCycle` which I think is inevitably here - the `TypeWalker` tracks what constitutes a cycle and so you need to have a hook to say: "When you find a cycle what do you want me to do" - there's no other place that bit of information could go I don't think.

Reply:
I think I follow what you're saying about pushing the before and after hooks down into the visitor, but it's hard to visualise a better way to do it without seeing the usage. Presumably the visitor that gets passed in when making the walker has already had some some custom visit functions applied? One argument for keeping them as TypeWalker configuration is if it would ever make sense to use a customised visitor with two different TypeWalkers with different impls for before/after visit and cycle handling? That might be something that isn't ever really done though.

I don't think this seems bad as it is - definitely ok to land like this and then tweak it as clean ups occur to us.

Reply:
@babbageclunk good point w.r.t. tweaking in the future if need be